Sming/Components/MultipartParser/multipart-parser.patch

```diff
@@ -36,39 +36,55 @@ index 981dabb..52e0200 100644
  enum state {
    s_uninitialized = 1,
    s_start,
-@@ -77,10 +64,10 @@ enum state {
-   s_end
+   s_start_boundary,
++  s_preamble,
++  s_preamble_almost_done,
+   s_header_field_start,
+   s_header_field,
+   s_headers_almost_done,
+@@ -74,44 +62,24 @@ enum state {
+   s_part_data_almost_end,
+   s_part_data_end,
+   s_part_data_final_hyphen,
+-  s_end
++  s_epilogue
  };
  
 -multipart_parser* multipart_parser_init
 -    (const char *boundary, const multipart_parser_settings* settings) {
-+multipart_parser_t* multipart_parser_init
-+    (const char *boundary, const multipart_parser_settings_t * settings) {
++void multipart_parser_init
++    (multipart_parser_t* p, const char *boundary, size_t boundary_length, const multipart_parser_settings_t * settings) {
  
 -  multipart_parser* p = malloc(sizeof(multipart_parser) +
-+  multipart_parser_t* p = malloc(sizeof(multipart_parser_t) +
-                                strlen(boundary) +
-                                strlen(boundary) + 9);
- 
-@@ -96,22 +83,22 @@ multipart_parser* multipart_parser_init
-   return p;
- }
- 
+-                               strlen(boundary) +
+-                               strlen(boundary) + 9);
+-
+-  strcpy(p->multipart_boundary, boundary);
+-  p->boundary_length = strlen(boundary);
+-  
+-  p->lookbehind = (p->multipart_boundary + p->boundary_length + 1);
++  p->multipart_boundary = boundary;
```

Comment:
@aemseemann This is C code. And in C you should use strcpy to copy strings. Here you are making a copy of a pointer that gets invalidated at the time when you actually have to access `multipart_boundary`.

Reply:
The intention *is* to copy the pointer and not the string content (to avoid an additional allocation).

It is the caller's responsibility to ensure that the content stays stays valid (and unchanged) throughout the lifetime of the `multipart_parser`. This is ensured, because the caller is the `MultipartParser` constructor and both `multipart_parser` and the `boundary` string are both members of `MultipartParser`.

Reply:
@aemseemann  can you try the following: Compile HttpServer_FirmwareUpload in debug mode for the Host
```bash
cd $SMING_HOME/../sample/HttpServer_FirmwareUpload
make dist-clean
make clean
make SMING_ARCH=Host ENABLE_GDB=1 ENABLE_LWIPDEBUG=1 ENABLE_MALLOC_COUNT=0
make flash
```

Once that finishes if you have valgrind run the application by calling
```bash
# setup your virtual network interface. see: https://sming.readthedocs.io/en/latest/_inc/Sming/Arch/Host/Components/lwip/index.html
make valgrind
```

At least on my system valgrind starts to report issues with the `multipart_boundary`. Example:
```
==23134== Invalid read of size 1
==23134==    at 0x809CC14: multipart_parser_execute (multipart_parser.c:112)
```

You can also run the application in a visual debugger and start monitoring the bytes starting at the memory address that `multipart_boundary` points to. As far as I can see the first bytes are overwritten.

Reply:
Can confirm the bug, thanks for the explanation!
I accidentally took the address from the `MultipartParser` constructor argument, which was shadowing the `boundary` member. This is fixed now.

However, valgrind discovered an issue with `TcpConnection::checkSelfFree`:
```
==8800== Invalid read of size 1
==8800==    at 0x12C374: checkSelfFree (TcpConnection.h:191)
==8800==    by 0x12C374: TcpConnection::checkSelfFree() (TcpConnection.h:189)
==8800==    by 0x12CD26: TcpConnection::internalOnPoll() (TcpConnection.cpp:507)
```
The reason is in TcpConnection.cpp:500pp
``` cpp
err_t TcpConnection::internalOnPoll()
{
	//if (tcp->state != ESTABLISHED)
	//	return ERR_OK;

	sleep++;
	err_t res = onPoll();
	checkSelfFree(); /// <<< instance was deleted inside `onPoll`
	debug_tcp_ext("<poll");
	return res;
}
```
Could be fixed with:
```cpp
	err_t res = onPoll(); 
	if (res != ERR_TIMEOUT) {
		// on ERR_TIMEOUT, checkSelfFree is already called inside checkSelfFree
		checkSelfFree();
	}
```
However, I lack the in-depth knowledge of the low level classes and relying on the error code seems quite fragile.